### ISSUE TITLE
Dictionary throws with KeyNotFoundException with multiple, similar objects

### ISSUE THREAD

**Issue by nightblade9** [nightblade9/haxesharp#10]
Hi @bsinky,

While trying to consume a `Dictionary` in one of my games, I ran into an interesting problem where I get a `KeyNotFound` exception.  Here's a simplified test that reproduces the result:

```
    @Test
    public function getGetsValuesSetWithObjectsAsKeys()
    {
        var m1 = new Monster(7);
        var m2 = new Monster(3);
        var m3 = new Monster(3);
        var m4 = new Monster(3);

        var dictionary = new Dictionary<Monster, Int>();
        dictionary[m1] = 0;
        dictionary[m2] = 0;
        dictionary[m3] = 0;
        dictionary[m4] = 0;
        
        Assert.that(dictionary[m1], Is.equalTo(0));
        Assert.that(dictionary[m2], Is.equalTo(0));
        Assert.that(dictionary[m3], Is.equalTo(0));
        Assert.that(dictionary[m4], Is.equalTo(0));
    }
```

Helper class:

```
class Monster
{
    private var value:Int;

    public function new(value:Int)
    {
        this.value = value;
    }
}
```

Exception stack:

```
    ERROR: massive.munit.UnhandledException: The given key ({ value => 7 }) was not present in the Dictionary. Keys: { next => #function:0, a => [{ value => 7 },{ value => 3 },{ value => 3 },{ value => 3 }], p => 0, hasNext => #function:0 }
from: haxesharp.exceptions.KeyNotFoundException.new() at 19


Called from /usr/share/haxe/std/haxe/ds/BalancedTree.hx line 146
Called from /usr/share/haxe/std/haxe/ds/BalancedTree.hx line 150
Called from /usr/share/haxe/std/haxe/ds/BalancedTree.hx line 84 at haxesharp.collections.DictionaryTest#getGetsValuesSetWithObjectsAsKeys at haxesharp/collections/Dictionary.hx (41)
        at haxesharp/collections/DictionaryTest.hx (186)
        at massive/munit/TestRunner.hx (330)
```

(I also changed the default exception to print out the key and valid key values.)

Is this something you can look into?

**Comment by bsinky** [comment]
Looks like this is one of the reasons BalancedTree doesn't work as Dictionary's backing collection, it uses `Reflect.compare` for equality comparison, which only produces defined results for numeric types and string - for all other types, the result is undefined: http://api.haxe.org/Reflect.html#compare

I've been thinking on it a little, still not sure how to get around the issues with not being able to have Map<K,V> as the backing Dictionary value. :face_with_head_bandage: K and V aren't available at runtime at all, which makes me wonder if we could define a macro that switched on K's type, since it would be available at compile time, and return an expression that returns a new IntMap/StringMap/ObjectMap/EnumMap <V>.  Because from my tests, the only issue was with `_map = new Map<K,V>()` in the code.  The other branch of the constructor, where `?map` was not null, compiles fine from what I remember.

My Haxe macro knowledge is still virtually nonexistent though.

**Comment by nightblade9** [comment]
> My Haxe macro knowledge is still virtually nonexistent though.

Me too. Macros are not an option.

I see two options:

## Back with a Map<ObjectKey, Any>

The main storage is a `Map<ObjectKey, Any>`. To convert non-objects into objects, `ObjectKey` takes an int/enum/string/obj and wraps it into an object. 

To solve the problem of `d.set(1, x); d.get(1)` equality with object keys, we will add a second set of maps that map things (eg. ints) to `ObjectKey` instances. This will keep all operations O(1) (two lookups).

## Back with an Array

The other option is to just use a list/array as the back-end. To keep track of where things are in O(1), we keep four dictionaries internally; `Map<Int, Int>`, `Map<String, Int>`, `Map<Enum, Int>`, and `Map<Object, Int>`. The values are the index arrays, and then we can get/set the value in O(1).

The only issue I see is that we would always auto-increment the index, so even if you `remove` stuff, we never go back and put something into that array index. This means we use a bunch more space than necessary.

-----

Which of these two solutions seems better to you? They're similar in terms of code and runtime complexity for the most part.

**Comment by bsinky** [comment]
I would take ObjectMap<ObjectKey,Any>.  With an Array, the auto-incrementing index resulting in more and more storage requirements makes that seem like a non-option, especially for any long-lived Dictionaries that are frequently added to or removed from.  The potential performance penalty of ObjectKey seems much better.

Does this seem like the better of the 2 options to you?

That said, I think you should do this implementation for Dictionary, since you already have a solid plan for it in place.

**Comment by nightblade9** [comment]
Yes, I agree. The non-integer-reuse part seems very shady.

I'll take a crack at this, based on [my old prototype](https://try.haxe.org/#dDE77).

**Comment by nightblade9** [comment]
Looks like my plan didn't work. It depends on an O(1) mapping of `Any => ObjectWrapper`, which is impossible with the existing classes. There's no way to figure out if a key is in the dictionary without iterating all keys one way or another.

Since the generic type `T` only exists at compile time, I'm going to look into (possibly flaky) runtime ways to tell *something* useful about the class.

**Comment by nightblade9** [comment]
I got [this solution working](https://try.haxe.org/#16Cf1). It basically uses `Type.typeOf` and `Type.getClass` to reliably determine the type at runtime, and then stuffs the object into one of four dictionaries appropriately.

Elegant? No. Optimal? No. Performant? Yes, probably. 

Thumbs up or thumbs down?

**Comment by bsinky** [comment]
Working and performant get my 👍

**Comment by nightblade9** [comment]
I'll try to clean this up and submit a PR for it today.

The only drawback I see is: am I using some internal thing that may change in the future? But I think it's unlikely to change/break -- these methods seem stable to me.

Let's try this. Because frankly, I don't see any other option at this point.

**Comment by nightblade9** [comment]
Resolved with #12.
